### PR TITLE
[ONNX FE] Support training mode of BatchNormalization

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - ONNX `BatchNormalization` in the training mode was rejected by the ONNX frontend for every supported opset. It is now converted to a subgraph which normalizes the input with the statistics calculated for the current batch and produces the updated running statistics:
   - `running_mean = input_mean * momentum + current_mean * (1 - momentum)`
   - `running_var = input_var * momentum + current_var * (1 - momentum)`
 - Training mode is detected per opset: `is_test` attribute (opsets 1-6), number of requested outputs (opsets 7-13), `training_mode` attribute (opsets 14+). Inference mode behavior is unchanged.
 - Dynamic-rank inputs are supported (reduction axes are calculated in the runtime).
 - Added ONNX FE tests for opsets 1, 9, 14 and for a dynamic-rank model; removed the `xfail_issue_63033` marks from `test_batchnorm_example_training_mode` / `test_batchnorm_epsilon_training_mode` ONNX backend tests.

### Tickets:
 - CVS-192056
 - CVS-63033

### AI Assistance:
 - *AI assistance used: yes*
 - AI was used to implement the conversion and the tests. Human validation: the ONNX frontend unit test suite was built and executed locally (`ov_onnx_frontend_tests`, 1157 passed / 0 failed), the new training-mode conversion was additionally validated against the reference input/output data of the ONNX node tests `test_batchnorm_example_training_mode` and `test_batchnorm_epsilon_training_mode`, and `clang-format` was checked on the changed sources.

### CODE REVIEW

File: src/frontends/onnx/frontend/src/op/batch_norm.cpp [hunk 1]
```diff
@@ -10,45 +10,120 @@
...
+                                           double momentum) {
+    const auto axes = statistics_axes(x);
+    const auto mean_keep_dims = std::make_shared<v1::ReduceMean>(x, axes, true);
+    const auto deviation = std::make_shared<v1::Subtract>(x, mean_keep_dims);
+    const auto var_keep_dims =
+        std::make_shared<v1::ReduceMean>(std::make_shared<v1::Multiply>(deviation, deviation), axes, true);
+
+    const auto channels_shape = v0::Constant::create(ov::element::i64, ov::Shape{1}, {-1});
+    const ov::Output<ov::Node> current_mean = std::make_shared<v1::Reshape>(mean_keep_dims, channels_shape, false);
+    const ov::Output<ov::Node> current_var = std::make_shared<v1::Reshape>(var_keep_dims, channels_shape, false);
```
[the remark this answers is not in the document]
Comment (mvafin) [reply src/frontends/onnx/frontend/src/op/batch_norm.cpp]: Good catch, fixed. The batch statistics are now accumulated in `f32` whenever the input is a lower-precision float type, and converted back to the input precision only where `BatchNormInference` requires all inputs to share the element type. The running statistics are updated from the higher precision values. Added `onnx_model_batch_norm_training_mode_f16` which asserts the `ReduceMean` nodes run in `f32` (a pure runtime check would not catch it, because the template backend upcasts `f16` internally in `ReduceMean::evaluate`).

File: src/frontends/onnx/frontend/src/op/batch_norm.cpp [hunk 2]
```diff
@@ -10,45 +10,134 @@
...
     auto bias = inputs.at(2);
-    ov::Output<ov::Node> mean;
-    ov::Output<ov::Node> var;
 
-    double epsilon{node.get_attribute_value<double>("epsilon", 1e-5)};
+    CHECK_VALID_NODE(node,
+                     inputs.size() >= 5,
+                     "Cannot create OpenVINO batch norm with unsupported number of inputs: ",
+                     inputs.size());
+    auto mean = inputs.at(3);
```
Comment (bumbosiepsak) [inline src/frontends/onnx/frontend/src/op/batch_norm.cpp:122]: opset_1::batch_norm now guards the required inputs, but opset_7 (and opset_14) still read inputs 3/4 with no guard.

For an untrusted model with fewer than 5 inputs, these paths throw a raw std::out_of_range instead of a descriptive frontend error. Frontend inputs should fail fast with a clear message. Suggest adding the same CHECK_VALID_NODE(node, inputs.size() >= 5, ...) guard in opset_7 and opset_14 before reading inputs.at(3)/inputs.at(4).

File: src/frontends/onnx/frontend/src/op/batch_norm.cpp [hunk 3]
```diff
@@ -99,12 +192,19 @@ ov::OutputVector batch_norm(const ov::frontend::onnx::Node& node) {
...
+    const bool training_mode = node.get_attribute_value<int64_t>("training_mode", 0) != 0;
 
     CHECK_VALID_NODE(node,
-                     training_mode == false && node.get_outputs_size() == 1,
-                     "Training mode of BatchNormalization is not supported.");
-    return {std::make_shared<v5::BatchNormInference>(x, scale, bias, mean, var, epsilon)};
+                     training_mode || node.get_outputs_size() == 1,
+                     "Number of outputs greater than one requires the training mode to be enabled.");
+
+    if (!training_mode) {
```
Comment (bumbosiepsak) [inline src/frontends/onnx/frontend/src/op/batch_norm.cpp:212]: opset_14 emits 3 outputs even when only 1 is requested. When training_mode is true the code unconditionally returns 3 outputs regardless of node.get_outputs_size(). A model that sets training_mode=1 but declares a single output would get more outputs than declared. Consider gating the extra outputs on node.get_outputs_size() to avoid an output-count mismatch.

File: src/frontends/onnx/frontend/src/op/batch_norm.cpp
Comment (bumbosiepsak) [inline src/frontends/onnx/frontend/src/op/batch_norm.cpp]: Stale comment references a non-existent ReduceVar. The implementation computes variance via ReduceMean of the squared deviation — there is no ReduceVar op. Minor wording fix to avoid confusing future readers (e.g. "the mean and variance reductions use float").

File: src/frontends/onnx/frontend/src/op/batch_norm.cpp (hunk 2 above)
Comment (mvafin) [reply src/frontends/onnx/frontend/src/op/batch_norm.cpp:122]: Fixed: added the same `CHECK_VALID_NODE(node, inputs.size() >= 5, ...)` guard to `opset_7` and `opset_14`, and added a regression test (`onnx_model_batch_norm_opset7_invalid_inputs_number`) verifying a descriptive frontend error is thrown instead of `std::out_of_range`.

File: src/frontends/onnx/frontend/src/op/batch_norm.cpp (hunk 3 above)
Comment (mvafin) [reply src/frontends/onnx/frontend/src/op/batch_norm.cpp:212]: Fixed: `opset_14` now returns just `{bn.y}` when `node.get_outputs_size() == 1`, and only returns the 3 outputs (`y`, `running_mean`, `running_var`) otherwise. Added a regression test (`onnx_model_batch_norm_training_mode_opset14_single_output`) covering `training_mode=1` with a single declared output.

File: src/frontends/onnx/frontend/src/op/batch_norm.cpp
Comment (mvafin) [reply src/frontends/onnx/frontend/src/op/batch_norm.cpp]: Fixed: reworded both trailing comments (opset 6 and opset 15 sections) to "the mean and variance reductions use float" instead of referencing the non-existent `ReduceVar` op.

**Review by bumbosiepsak — APPROVED** [review]
